src/basic/hashmap.c

```diff
@@ -285,7 +286,10 @@ _destructor_ static void cleanup_pools(void) {
         /* The pool is only allocated by the main thread, but the memory can
          * be passed to other threads. Let's clean up if we are the main thread
          * and no other threads are live. */
-        if (!is_main_thread())
+        /* We can't use is_main_thread() here, as it uses a thread-local storage
+         * which, in turn, uses malloc() (at least in C11), which glibc/valgrind
```

Comment:
comment is slightly misleading... the thread local storage construct we use is C11 is specific, and there is no version of it on older C versions, and that means also none which use malloc() internally.

Maybe this:

> We build our own is_main_thread() here, which doesn't use C11 TLS based caching of the result. That's because valgrind apparently doesn't like malloc() (which C11 TLS internally uses) to be called from a GCC destructors.

Reply:
That sounds much better, thanks!